Fix built-in `result` variable for optional resource typed returns

SupunS opened the pull request:
Closes #2443

## Description

Make the `result` variable an optional-reference, instead of a reference to an optional.

______

<!-- Complete: -->

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

turbolent reviewed (commented):
Nice, good catch!

I guess this is mostly a UX issue where currently the result cannot be accessed

This logic should have probably changed back when we changed from reference-to-optional to optional-reference (e.g. in reference expressions).

Should this maybe be recursively defined, instead of just unwrapping one level? For example, given a (unusual) result type `R??` (where `R` is resource-kinded), we probably want to get `(&R)??`.

Also, I'm not quite sure, but we might have missed the resource-kinded reference tracking for this reference value creation here and should add it in case it is missing. We want to maybe also double check if there are other cases where we create references to resource-kinded values and are missing the tracking.

SupunS commented:
> This logic should have probably changed back when we changed from reference-to-optional to optional-reference (e.g. in > reference expressions).
> 
> Should this maybe be recursively defined, instead of just unwrapping one level? For example, given a (unusual) result type > R?? (where R is resource-kinded), we probably want to get (&R)??.

Yeah, I changed this to be aligned with the optional reference change (https://github.com/onflow/cadence/pull/1303). There also we only unwrap it just one level (see below). Not sure if we need to do the same for nested optionals as well. Need to think about it a bit more. But I think that should probably be a separate change (and maybe breaking).

https://github.com/onflow/cadence/blob/a87f3634b5d760caa20d28e9be3f95332374ea30/runtime/sema/check_reference_expression.go#L45-L50

turbolent commented:
As discussed earlier, agreed on aligning with the behaviour of reference expressions, i.e. not unwrapping nested optionals, leaving the current PR's behaviour 👍 

Though maybe unlikely to be a real problem currently, it would still be great to add the reference tracking sure for consistency reasons here, and maybe elsewhere where missing. It might be good to do that in a separate PR.
